File: frontend/src/services/creditTransactionService.ts
/**
 * Credit Transaction Service - API calls for credit transactions
 */

const API_BASE_URL = '/api/credit-transactions';

export interface QuickCreditPickupRequest {
  customerCode: string;
  items: Array<{
    productCode: string;
    quantity: number;
    notes?: string;
  }>;
  performedBy: string;
  performedByRole: string;
  notes?: string;
}

export interface CreditTransactionResponse {
  id: string;
  transactionNumber: string;
  customerId: string;
  customerCode: string;
  customerName: string;
  transactionType: 'PICKUP' | 'RETURN';
  status: 'PENDING' | 'CONFIRMED' | 'INVOICED' | 'CANCELLED';
  lines: Array<{
    id: string;
    productId: string;
    productCode: string;
    productName: string;
    quantity: number;
    unitPrice: number;
    lineTotal: number;
    notes?: string;
  }>;
  totalAmount: number;
  totalItems: number;
  performedBy: string;
  performedByRole: string;
  confirmedBy?: string;
  confirmedAt?: string;
  notes?: string;
  createdAt: string;
  updatedAt: string;
}

export interface CreditTransactionSummaryResponse {
  id: string;
  transactionNumber: string;
  customerCode: string;
  customerName: string;
  transactionType: 'PICKUP' | 'RETURN';
  status: 'PENDING' | 'CONFIRMED' | 'INVOICED' | 'CANCELLED';
  totalAmount: number;
  totalItems: number;
  performedBy: string;
  createdAt: string;
}

/**
 * Create quick credit pickup
 */
export const createQuickCreditPickup = async (
  request: QuickCreditPickupRequest
): Promise<CreditTransactionResponse> => {
  const response = await fetch(`${API_BASE_URL}/quick-pickup`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(request),
  });

  if (!response.ok) {
    throw new Error('Failed to create credit pickup');
  }

  return response.json();
};

/**
 * Get all credit transactions
 */
export const getAllCreditTransactions = async (
  page: number = 0,
  size: number = 20
): Promise<{ content: CreditTransactionSummaryResponse[]; totalElements: number }> => {
  const response = await fetch(`${API_BASE_URL}?page=${page}&size=${size}`);

  if (!response.ok) {
    throw new Error('Failed to fetch credit transactions');
  }

  return response.json();
};

/**
 * Get credit transaction by ID
 */
export const getCreditTransactionById = async (
  id: string
): Promise<CreditTransactionResponse> => {
  const response = await fetch(`${API_BASE_URL}/${id}`);

  if (!response.ok) {
    throw new Error('Failed to fetch credit transaction');
  }

  return response.json();
};

/**
 * Get credit transactions for a customer
 */
export const getCustomerCreditTransactions = async (
  customerId: string,
  page: number = 0,
  size: number = 20
): Promise<{ content: CreditTransactionSummaryResponse[]; totalElements: number }> => {
  const response = await fetch(`${API_BASE_URL}/customer/${customerId}?page=${page}&size=${size}`);

  if (!response.ok) {
    throw new Error('Failed to fetch customer credit transactions');
  }

  return response.json();
};

/**
 * Get recent credit transactions for a customer
 */
export const getRecentCustomerCreditTransactions = async (
  customerId: string,
  limit: number = 10
): Promise<CreditTransactionSummaryResponse[]> => {
  const response = await fetch(`${API_BASE_URL}/customer/${customerId}/recent?limit=${limit}`);

  if (!response.ok) {
    throw new Error('Failed to fetch recent credit transactions');
  }

  return response.json();
};

/**
 * Get pending credit transactions for a customer
 */
export const getPendingCustomerCreditTransactions = async (
  customerId: string
): Promise<CreditTransactionResponse[]> => {
  const response = await fetch(`${API_BASE_URL}/customer/${customerId}/pending`);

  if (!response.ok) {
    throw new Error('Failed to fetch pending credit transactions');
  }

  return response.json();
};

/**
 * Search credit transactions
 */
export const searchCreditTransactions = async (
  query: string,
  page: number = 0,
  size: number = 20
): Promise<{ content: CreditTransactionSummaryResponse[]; totalElements: number }> => {
  const response = await fetch(`${API_BASE_URL}/search?q=${encodeURIComponent(query)}&page=${page}&size=${size}`);

  if (!response.ok) {
    throw new Error('Failed to search credit transactions');
  }

  return response.json();
};

/**
 * Confirm credit transaction
 */
export const confirmCreditTransaction = async (
  id: string,
  confirmedBy: string,
  signatureData?: string,
  notes?: string
): Promise<CreditTransactionResponse> => {
  const response = await fetch(`${API_BASE_URL}/${id}/confirm`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      confirmedBy,
      signatureData,
      notes,
    }),
  });

  if (!response.ok) {
    throw new Error('Failed to confirm credit transaction');
  }

  return response.json();
};

/**
 * Cancel credit transaction
 */
export const cancelCreditTransaction = async (
  id: string,
  reason: string
): Promise<void> => {
  const response = await fetch(`${API_BASE_URL}/${id}/cancel?reason=${encodeURIComponent(reason)}`, {
    method: 'POST',
  });

  if (!response.ok) {
    throw new Error('Failed to cancel credit transaction');
  }
};

/**
 * Get monthly statement for customer
 */
export const getMonthlyStatement = async (
  customerId: string,
  year: number,
  month: number
): Promise<CreditTransactionResponse[]> => {
  const response = await fetch(`${API_BASE_URL}/customer/${customerId}/statement/${year}/${month}`);

  if (!response.ok) {
    throw new Error('Failed to fetch monthly statement');
  }

  return response.json();
};

/**
 * Generate monthly statement PDF (example)
 */
export const generateMonthlyStatementPDF = async (
  customerId: string,
  year: number,
  month: number
): Promise<Blob> => {
  const response = await fetch(`${API_BASE_URL}/customer/${customerId}/statement/${year}/${month}/pdf`);

  if (!response.ok) {
    throw new Error('Failed to generate monthly statement PDF');
  }

  return response.blob();
};

File: backend/src/main/java/lt/elektromeistras/service/CreditTransactionService.java
package lt.elektromeistras.service;

import lt.elektromeistras.domain.*;
import lt.elektromeistras.dto.request.*;
import lt.elektromeistras.dto.response.*;
import lt.elektromeistras.repository.*;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.Instant;
import java.time.LocalDate;
import java.time.ZoneId;
import java.time.temporal.TemporalAdjusters;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
@Transactional(readOnly = true)
public class CreditTransactionService {

    private final CreditTransactionRepository transactionRepository;
    private final CreditTransactionLineRepository lineRepository;
    private final CustomerRepository customerRepository;
    private final ProductRepository productRepository;

    /**
     * ULTRA FAST credit pickup - optimized for rapid entry
     * Uses customer code and product codes for instant lookup
     */
    @Transactional
    public CreditTransactionResponse createQuickCreditPickup(QuickCreditPickupRequest request) {
        log.info("Creating quick credit pickup for customer: {}", request.getCustomerCode());

        // 1. Find customer by code (FAST - indexed lookup)
        Customer customer = customerRepository.findByCode(request.getCustomerCode())
                .orElseThrow(() -> new RuntimeException("Customer not found: " + request.getCustomerCode()));

        // 2. Create transaction
        CreditTransaction transaction = new CreditTransaction();
        transaction.setCustomer(customer);
        transaction.setTransactionType(CreditTransaction.TransactionType.PICKUP);
        transaction.setStatus(CreditTransaction.TransactionStatus.PENDING);
        transaction.setPerformedBy(request.getPerformedBy());
        transaction.setPerformedByRole(CreditTransaction.PerformedByRole.valueOf(request.getPerformedByRole()));
        transaction.setNotes(request.getNotes());

        // 3. Add line items
        for (QuickCreditPickupRequest.QuickCreditLineItem item : request.getItems()) {
            Product product = productRepository.findByCode(item.getProductCode())
                    .orElseThrow(() -> new RuntimeException("Product not found: " + item.getProductCode()));

            CreditTransactionLine line = new CreditTransactionLine();
            line.setProduct(product);
            line.setProductCode(product.getCode());
            line.setProductName(product.getName());
            line.setQuantity(item.getQuantity());
            line.setUnitPrice(product.getBasePrice());
            line.setNotes(item.getNotes());

            transaction.addLine(line);
        }

        // 4. Save transaction
        CreditTransaction saved = transactionRepository.save(transaction);

        log.info("Quick credit pickup created: {} with {} items",
                saved.getTransactionNumber(), saved.getTotalItems());

        return mapToResponse(saved);
    }

    /**
     * Create credit transaction (pickup or return)
     */
    @Transactional
    public CreditTransactionResponse createTransaction(CreateCreditTransactionRequest request) {
        log.info("Creating credit transaction for customer: {}", request.getCustomerId());

        // 1. Find customer
        Customer customer = customerRepository.findById(request.getCustomerId())
                .orElseThrow(() -> new RuntimeException("Customer not found: " + request.getCustomerId()));

        // 2. Create transaction
        CreditTransaction transaction = new CreditTransaction();
        transaction.setCustomer(customer);
        transaction.setTransactionType(CreditTransaction.TransactionType.valueOf(request.getTransactionType()));
        transaction.setStatus(CreditTransaction.TransactionStatus.PENDING);
        transaction.setPerformedBy(request.getPerformedBy());
        transaction.setPerformedByRole(CreditTransaction.PerformedByRole.valueOf(request.getPerformedByRole()));
        transaction.setNotes(request.getNotes());

        // 3. Add line items
        for (CreditTransactionLineRequest lineReq : request.getLines()) {
            UUID productId = lineReq.getProductId();
            if (productId == null && lineReq.getProductCode() != null) {
                Product product = productRepository.findByCode(lineReq.getProductCode())
                        .orElseThrow(() -> new RuntimeException("Product not found: " + lineReq.getProductCode()));
                productId = product.getId();
            }

            Product product = productRepository.findById(productId)
                    .orElseThrow(() -> new RuntimeException("Product not found: " + productId));

            CreditTransactionLine line = new CreditTransactionLine();
            line.setProduct(product);
            line.setProductCode(product.getCode());
            line.setProductName(product.getName());
            line.setQuantity(lineReq.getQuantity());
            line.setUnitPrice(product.getBasePrice());
            line.setNotes(lineReq.getNotes());

            transaction.addLine(line);
        }

        // 4. Save transaction
        CreditTransaction saved = transactionRepository.save(transaction);

        log.info("Credit transaction created: {} with {} items",
                saved.getTransactionNumber(), saved.getTotalItems());

        return mapToResponse(saved);
    }

    /**
     * Confirm credit transaction and update customer balance
     */
    @Transactional
    public CreditTransactionResponse confirmTransaction(UUID transactionId, ConfirmCreditTransactionRequest request) {
        log.info("Confirming credit transaction: {}", transactionId);

        CreditTransaction transaction = transactionRepository.findById(transactionId)
                .orElseThrow(() -> new RuntimeException("Transaction not found: " + transactionId));

        if (transaction.getStatus() != CreditTransaction.TransactionStatus.PENDING) {
            throw new RuntimeException("Only pending transactions can be confirmed");
        }

        // Confirm transaction
        transaction.confirm(request.getConfirmedBy());
        if (request.getSignatureData() != null) {
            transaction.setSignatureData(request.getSignatureData());
        }
        if (request.getPhotoData() != null) {
            transaction.setPhotoData(request.getPhotoData());
        }
        if (request.getNotes() != null) {
            transaction.setNotes(transaction.getNotes() + "\n" + request.getNotes());
        }

        // Update customer balance
        Customer customer = transaction.getCustomer();
        if (transaction.getTransactionType() == CreditTransaction.TransactionType.PICKUP) {
            // Pickup increases balance (debt)
            customer.setCurrentBalance(customer.getCurrentBalance().add(transaction.getTotalAmount()));
        } else {
            // Return decreases balance (payment)
            customer.setCurrentBalance(customer.getCurrentBalance().subtract(transaction.getTotalAmount()));
        }
        customerRepository.save(customer);

        CreditTransaction saved = transactionRepository.save(transaction);

        log.info("Credit transaction confirmed: {} - Customer balance updated: {}",
                saved.getTransactionNumber(), customer.getCurrentBalance());

        return mapToResponse(saved);
    }

    /**
     * Cancel transaction
     */
    @Transactional
    public void cancelTransaction(UUID transactionId, String reason) {
        log.info("Cancelling credit transaction: {}", transactionId);

        CreditTransaction transaction = transactionRepository.findById(transactionId)
                .orElseThrow(() -> new RuntimeException("Transaction not found: " + transactionId));

        if (transaction.getStatus() == CreditTransaction.TransactionStatus.INVOICED) {
            throw new RuntimeException("Cannot cancel invoiced transactions");
        }

        // If already confirmed, reverse customer balance
        if (transaction.getStatus() == CreditTransaction.TransactionStatus.CONFIRMED) {
            Customer customer = transaction.getCustomer();
            if (transaction.getTransactionType() == CreditTransaction.TransactionType.PICKUP) {
                customer.setCurrentBalance(customer.getCurrentBalance().subtract(transaction.getTotalAmount()));
            } else {
                customer.setCurrentBalance(customer.getCurrentBalance().add(transaction.getTotalAmount()));
            }
            customerRepository.save(customer);
        }

        transaction.setStatus(CreditTransaction.TransactionStatus.CANCELLED);
        transaction.setNotes(transaction.getNotes() + "\nCANCELLED: " + reason);
        transactionRepository.save(transaction);

        log.info("Credit transaction cancelled: {}", transaction.getTransactionNumber());
    }

    /**
     * Get transaction by ID
     */
    public CreditTransactionResponse getTransactionById(UUID id) {
        CreditTransaction transaction = transactionRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Transaction not found: " + id));
        return mapToResponse(transaction);
    }

    /**
     * Get transaction by transaction number
     */
    public CreditTransactionResponse getTransactionByNumber(String transactionNumber) {
        CreditTransaction transaction = transactionRepository.findByTransactionNumber(transactionNumber)
                .orElseThrow(() -> new RuntimeException("Transaction not found: " + transactionNumber));
        return mapToResponse(transaction);
    }

    /**
     * Get all transactions for a customer
     */
    public Page<CreditTransactionSummaryResponse> getCustomerTransactions(UUID customerId, Pageable pageable) {
        Page<CreditTransaction> transactions = transactionRepository.findByCustomerId(customerId, pageable);
        return transactions.map(this::mapToSummaryResponse);
    }

    /**
     * Get recent transactions for a customer
     */
    public List<CreditTransactionSummaryResponse> getRecentCustomerTransactions(UUID customerId, int limit) {
        List<CreditTransaction> transactions = transactionRepository.findRecentByCustomer(
                customerId,
                PageRequest.of(0, limit, Sort.by(Sort.Direction.DESC, "createdAt"))
        );
        return transactions.stream()
                .map(this::mapToSummaryResponse)
                .collect(Collectors.toList());
    }

    /**
     * Get pending transactions for a customer
     */
    public List<CreditTransactionResponse> getPendingCustomerTransactions(UUID customerId) {
        List<CreditTransaction> transactions = transactionRepository.findPendingByCustomer(customerId);
        return transactions.stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    /**
     * Search transactions
     */
    public Page<CreditTransactionSummaryResponse> searchTransactions(String query, Pageable pageable) {
        Page<CreditTransaction> transactions = transactionRepository.searchTransactions(query, pageable);
        return transactions.map(this::mapToSummaryResponse);
    }

    /**
     * Get all transactions
     */
    public Page<CreditTransactionSummaryResponse> getAllTransactions(Pageable pageable) {
        Page<CreditTransaction> transactions = transactionRepository.findAll(pageable);
        return transactions.map(this::mapToSummaryResponse);
    }

    /**
     * Get monthly statement for customer
     * Returns all confirmed transactions for the specified month
     */
    public List<CreditTransactionResponse> getMonthlyStatement(UUID customerId, int year, int month) {
        log.info("Generating monthly statement for customer: {} - {}/{}", customerId, year, month);

        LocalDate startOfMonth = LocalDate.of(year, month, 1);
        LocalDate endOfMonth = startOfMonth.with(TemporalAdjusters.lastDayOfMonth());

        Instant startDate = startOfMonth.atStartOfDay(ZoneId.systemDefault()).toInstant();
        Instant endDate = endOfMonth.plusDays(1).atStartOfDay(ZoneId.systemDefault()).toInstant();

        List<CreditTransaction> transactions = transactionRepository.findByCustomerAndDateRange(
                customerId, startDate, endDate
        );

        return transactions.stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    /**
     * Map to full response
     */
    private CreditTransactionResponse mapToResponse(CreditTransaction t) {
        CreditTransactionResponse response = new CreditTransactionResponse();
        response.setId(t.getId());
        response.setTransactionNumber(t.getTransactionNumber());
        response.setCustomerId(t.getCustomer().getId());
        response.setCustomerCode(t.getCustomer().getCode());
        response.setCustomerName(t.getCustomer().getDisplayName());
        response.setTransactionType(t.getTransactionType().name());
        response.setStatus(t.getStatus().name());
        response.setTotalAmount(t.getTotalAmount());
        response.setTotalItems(t.getTotalItems());
        response.setPerformedBy(t.getPerformedBy());
        response.setPerformedByRole(t.getPerformedByRole().name());
        response.setConfirmedBy(t.getConfirmedBy());
        response.setConfirmedAt(t.getConfirmedAt());
        response.setSignatureData(t.getSignatureData());
        response.setPhotoData(t.getPhotoData());
        response.setNotes(t.getNotes());
        response.setCreatedAt(t.getCreatedAt());
        response.setUpdatedAt(t.getUpdatedAt());

        // Map lines
        List<CreditTransactionLineResponse> lineResponses = t.getLines().stream()
                .map(this::mapLineToResponse)
                .collect(Collectors.toList());
        response.setLines(lineResponses);

        return response;
    }

    /**
     * Map to summary response
     */
    private CreditTransactionSummaryResponse mapToSummaryResponse(CreditTransaction t) {
        return new CreditTransactionSummaryResponse(
                t.getId(),
                t.getTransactionNumber(),
                t.getCustomer().getCode(),
                t.getCustomer().getDisplayName(),
                t.getTransactionType().name(),
                t.getStatus().name(),
                t.getTotalAmount(),
                t.getTotalItems(),
                t.getPerformedBy(),
                t.getCreatedAt()
        );
    }

    /**
     * Map line to response
     */
    private CreditTransactionLineResponse mapLineToResponse(CreditTransactionLine line) {
        return new CreditTransactionLineResponse(
                line.getId(),
                line.getProduct().getId(),
                line.getProductCode(),
                line.getProductName(),
                line.getQuantity(),
                line.getUnitPrice(),
                line.getLineTotal(),
                line.getNotes()
        );
    }
}
